packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,652 @@
+// We allow dead-code here, because some of the functions are only exposed when `feature = "hooks"`
+// we could do cfg for everything, but that gets very messy, instead we only use a subset
+// and enable deadcode on `feature = "std"`.
+#![cfg_attr(not(feature = "std"), allow(dead_code))]
+
+use alloc::{boxed::Box, collections::BTreeMap, vec::Vec};
+use core::{
+    any::{Any, TypeId},
+    marker::PhantomData,
+};
+
+pub use default::builtin;
+
+use crate::fmt::{Emit, Frame, Snippet};
+
+#[derive(Default)]
+pub struct HookContextImpl {
+    pub(crate) snippets: Vec<Snippet>,
+    alternate: bool,
+
+    storage: BTreeMap<TypeId, BTreeMap<TypeId, Box<dyn Any>>>,
+}
+
+impl HookContextImpl {
+    pub(crate) fn cast<T>(&mut self) -> HookContext<T> {
+        HookContext {
+            parent: self,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub(crate) const fn alternate(&self) -> bool {
+        self.alternate
+    }
```

Comment:
Do we have other information from `Formatter`, which are useful here? Ideally we would have access to the `Formatter`, but I think this is not easily possible because of the different `Emit` variants I guess?

Reply:
That's correct, we need the ability to reorder, indent and manipulate the stream, therefore passing the `Formatter` directly isn't really feasible.

Reply:
We could, in theory, expose:
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.fill
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.align
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.width
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.precision
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.sign_plus
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.sign_minus
* https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.sign_aware_zero_pad

But I don't think any of those are useful in the hook context. @TimDiekmann, what do you think?

Reply:
No, I think `alternate` is the most important one here